sudo-rs: init at 0.2.0

nicoonoclaste opened the pull request:
## Description of changes

- [x] packaged `sudo-rs`, a pure-Rust, mostly-compatible implementation of sudo
- [ ] *didn't* work (in this PRs) on the NixOS modules to allow users to switch to it

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
- [x] Tested, as applicable:
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

nicoonoclaste committed:
sudo-rs: init at 0.2.0

Closes: #252193

Nyabinary commented:
Damn

colemickens commented:
this partially addresses: https://github.com/NixOS/nixpkgs/issues/252193 (just to link these1)

R-VdP commented:
@Nyabinary: you'll need [a patch like this](https://github.com/R-VdP/sudo-rs/commit/a44541dcb36b94f938daaed66b3ff06cfc1c2b40) to make it work with our security wrappers. I'll be submitting this upstream but I need some time to do a proper write-up and finish the testing.

You'll also need to create an additional PAM service, called `sudo-i` for login shells using sudo, see [here](https://github.com/memorysafety/sudo-rs/blob/9a7f38fbddc59f40a5e0a57555131b36e09811e4/src/sudo/pam.rs#L114).

And we need to disable a bunch of stuff in the sudoers file that is not supported yet, things like `SETENV` and the user-specific `Defaults` that are included by the terminfo module.

This will need some testing as well. I think it would be nice to either extend or semi-duplicate the existing sudo nixos test.

Nyabinary commented:
![image](https://github.com/NixOS/nixpkgs/assets/97130632/8b5cc832-4fb1-47c3-b17e-e972f73c2696)
options like this should be added too right?

R-VdP commented:
> ![image](https://github.com/NixOS/nixpkgs/assets/97130632/8b5cc832-4fb1-47c3-b17e-e972f73c2696)
> Options like this should be added too right?

Not necessarily, as sudo-rs aims to be a drop-in replacement for sudo. So we could just set `security.sudo.package = pkgs.sudo-rs`.
This is not currently possible yet though, since many things are not supported yet by sudo-rs.

nicoonoclaste committed:
nixos/sudo: Guard against `security.sudo.package = pkgs.sudo-rs;`

This is not unlikely to happen, given the enthusiasm shown by some users,
but we are not there yet, and this will save them from breaking their system.

nicoonoclaste committed:
sudo-rs: Patch to work with NixOS' suid wrappers

happysalada commented on pkgs/tools/security/sudo-rs/default.nix:
```diff
@@ -0,0 +1,81 @@
+  pname = "sudo-rs";
+  version = "0.2.0";
+
+  src = fetchCrate {
```
is there a reason you chose fetchCrate instead of fetchFromGitHub ?
the github fetch should make the upstream code easier to inspect.

happysalada replied:
fetchCrate is definitely harder for anyone to inspect, but if you don't feel like doing it, it can be done sometimes later or never, I don't think it's a big deal.

happysalada commented on pkgs/tools/security/sudo-rs/default.nix:
```diff
@@ -0,0 +1,81 @@
+    changelog = "${meta.homepage}/blob/v${version}/CHANGELOG.md";
+    license = with licenses; [ asl20 mit ];
+    maintainers = with maintainers; [ nicoo ];
+    platforms = platforms.linux;
```
should this just be platforms.unix here ?
I don't see why this wouldn't work with darwin (of course there might need to be some adaptations made to the package, but leaving it built until someone figures out the details doesn't seem crazy to me).

happysalada commented on pkgs/tools/security/sudo-rs/default.nix:
```diff
@@ -0,0 +1,81 @@
+  meta = with lib; {
+    description = "A memory safe implementation of sudo and su.";
+    homepage = "https://github.com/memorysafety/sudo-rs";
+    changelog = "${meta.homepage}/blob/v${version}/CHANGELOG.md";
```
```suggestion
    changelog = "https://github.com/memorysafety/sudo-rs/releases/tag/v${version}";
```

happysalada commented on pkgs/tools/security/sudo-rs/default.nix:
```diff
@@ -0,0 +1,81 @@
+  postPatch = ''
+    substituteInPlace util/generate-docs.sh \
+      --replace "/usr/bin/env bash" ${lib.getExe bash} \
+      --replace util/pandoc.sh pandoc
```
why not also use the lib.getExe trick here too ?

happysalada commented:
Would you mind squashing the 3 commits into 1 ? (I don't see a strong reason to have 3 commits here).

Nyabinary commented:
Any update on this?

happysalada commented:
I'm waiting for ofborg and will squash the 3 commits into 1.

happysalada commented:
the 3 commits are just part of adding this package.
the sudo module's change is just a warning, IMO, this is just part of initializing the package.

mkg20001 commented:
this is sudo-rs. maybe future sudo. history and package file will be under intense scrutiny and people will want to know why what was done and will ask git blame. this is one of the packages where i'd just argue for no just because.

happysalada commented:
alright let's get a second opinion then.
@SuperSandro2000 sorry to bother, but just in case you have time, what do you think about having this PR in 1 commit or 3 commits ?

happysalada commented:
I was trying to find someone that might have time to help on this.
I am not going to take this issue to the grave though, I'll leave it 3 commits. It's definitely not worth arguing over.
I'll wait until ofborg finishes.

Nyabinary commented:
Hmm, what are the options for this package? Well, how to use it over sudo?

Nyabinary commented:
@mkg20001 Any PR for the progress so far?

mkg20001 commented:
@Nyabinary its that plus more commits

mkg20001 commented:
btw lets move further discussion to the tracking issue: https://github.com/NixOS/nixpkgs/issues/253465

R-VdP commented:
@Nyabinary feel free to help us out and propose changes that will help move this forward.

However, simply asking about the progress every other day is frankly a bit annoying and doesn't add anything.

So either be patient, or put in the time and effort yourself.
